sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullFunctions.scala

```diff
@@ -210,14 +210,58 @@ case class IsNotNull(child: Expression) extends UnaryExpression with Predicate {
   }
 }
 
+/**
+ * A predicate that is evaluated to be true if there are at least `n` null values.
+ */
+case class AtLeastNNulls(n: Int, children: Seq[Expression]) extends Predicate {
```

Comment:
Just to briefly clarify, I guess that the problem was that `AtLeastNNulls` also dropped `NaNs` but that we can't do that since it would lead to a violation of our NaN-equality semantics when joining on float/double columns?